File: DSA/LC/39_combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        combination = []

        def rec(index, total):
            if total > target or index >= len(nums):
                if total == target:
                    res.append(combination.copy())
                return None

            combination.append(nums[index])
            rec(index, total + nums[index])
            combination.pop()

            rec(index + 1, total)

        rec(0, 0)
        return res

File: DSA/LC/39_combination_sum.cpp
#include <vector>
#include <functional>

class Solution
{
public:
  std::vector<std::vector<int>> combinationSum(std::vector<int> &candidates, int target)
  {
    std::vector<std::vector<int>> res{};

    std::vector<int> combination{};

    const std::function<void(const int, const int)> rec = [&](const int index, const int sum)
    {
      if (index >= candidates.size() || sum > target)
      {
        if (sum == target)
          res.push_back(combination);
        return;
      }

      combination.push_back(candidates[index]);
      rec(index, sum + candidates[index]);
      combination.pop_back();

      rec(index + 1, sum);
    };

    rec(0, 0);
    return res;
  }
};